### PR TITLE
MDCT-2025: DynamicField Deletion Deletes Related Data

### PR DESCRIPTION
**Pull request by karla-vm**
## Description

Closes [MDCT-2025](https://qmacbis.atlassian.net/browse/MDCT-2025).

`DynamicField` (i.e. plans or BSS entities) have related sanctions and quality measures responses. As it stands, these sanctions and quality measures responses aren't deleted when an entity is deleted, causing data mismatching in the DB.

With the changes in this PR, the associated sanctions and quality measures responses are now deleted at the same time the entity is deleted.

Changes:
- Modifies entity deletion method to include filtering sanction and quality measures data before setting new entity, sanction, and quality measure data all at once.
- Adds associated tests.
- Modifies backend validation schema (frontend validation schema change not necessary) to allow entities (plans, bss entities) to be set to `[]` so we can successfully delete all entities and return to initial state (fixes a previous bug).
- Minor tweaks related to ^^ to prevent console errors.

### How to test

1. Add some plans
    - Navigate to `/mcpar/program-information/add-plans`
    - Add two plans (mentally select one to delete and remember it for the next step)
2. Add some sanctions and quality measures
    - Navigate to `/mcpar/plan-level-indicators/sanctions`
    - Create a sanction (pick the plan you intend to delete) -- no need to fill out the data in the drawer
    - Navigate to `/mcpar/plan-level-indicators/quality-measures`
    - Create a quality measure and enter data in the drawer for each plan measure
3. Delete the plan and check
    - Navigate to `/mcpar/program-information/add-plans`
    - Delete the plan you selected earlier
    - Navigate to `/mcpar/plan-level-indicators/sanctions`. The sanction should be gone!
    - Navigate to `/mcpar/plan-level-indicators/quality-measures` and open the drawer. The response for the plan you delete should be gone!

### Changed Dependencies
N/A

## Code author checklist
- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://bit.ly/3zPrxuZ) tests
- [x] ~~I have added analytics, if necessary~~
- [x] ~~I have updated the documentation, if necessary~~

## Reviewer checklist (two different people)
- [ ] I have done the deep review and verified the items in the above checklist are g2g
- [ ] I have done the lgtm review

### CODE REVIEW

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
-import { DeleteDynamicFieldRecordModal } from "components";
+import { DeleteDynamicFieldRecordModal, ReportContext } from "components";
 import { svgFilters } from "styles/theme";
 // utils
-import { EntityShape, EntityType, InputChangeEvent } from "types";
+import { EntityShape, EntityType, InputChangeEvent, ReportStatus } from "types";
+import { useUser } from "utils";
 // assets
 import cancelIcon from "assets/icons/icon_cancel_x_circle.png";
+import { AnyObject } from "yup/lib/object";
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/DynamicField.tsx:14]: I think we want the "AnyObject" from types, right?

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 2]
```diff
@@ -54,6 +59,61 @@ export const DynamicField = ({ name, label, ...props }: Props) => {
...
+        state: state,
+        id: report?.id,
+      };
+
+      let filteredEntities = {};
+      if (name === "plans") {
+        filteredEntities = {
+          [name]: form
+            .getValues()
+            .plans.filter((plan: AnyObject) => plan.id !== selectedRecord.id),
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/DynamicField.tsx:75]: I think you can use `[name]` here too and avoid the if/elseif. Then maybe in the filter just use a generic name

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 3]
```diff
@@ -54,6 +59,61 @@ export const DynamicField = ({ name, label, ...props }: Props) => {
...
+      }
+
+      const dataToWrite = {
+        status: ReportStatus.IN_PROGRESS,
+        lastAlteredBy: full_name,
+        fieldData: filteredEntities,
+      };
+
+      // delete related sanctions and quality measures
+      if (report?.fieldData.sanctions) {
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/DynamicField.tsx:94]: does we need a question mark before `sanctions`? (does it throw an error if we delete a plan when there are no sanctions?)

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 4]
```diff
@@ -54,6 +59,61 @@ export const DynamicField = ({ name, label, ...props }: Props) => {
...
+        const filteredSanctions = {
+          ["sanctions"]: report.fieldData.sanctions.filter(
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/DynamicField.tsx:96]: does this need brackets and quotes or could it just be `sanctions`?

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 5]
```diff
@@ -54,6 +59,61 @@ export const DynamicField = ({ name, label, ...props }: Props) => {
...
+            (sanction: EntityShape) =>
+              sanction.sanction_planName.value !== selectedRecord.id
+          ),
+        };
+        const dataToWrite = {
+          status: ReportStatus.IN_PROGRESS,
+          lastAlteredBy: full_name,
+          fieldData: filteredSanctions,
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/DynamicField.tsx:104]: does this confilct/overwrite the plans update?

**Comment by gmrabian** [comment]
The api was acting a little funny:
throws a 500 when you delete the only plan (probably because it's trying to write 0 plans and that's failing validation).
I also saw a 500 when I deleted one plan that had a sanction, so maybe sanction validation had a problem?

**Comment by gmrabian** [comment]
The api was acting a little funny:
throws a 500 when you delete the only plan (probably because it's trying to write 0 plans and that's failing validation).
I also saw a 500 when I deleted one plan that had a sanction, so maybe sanction validation had a problem?

**Comment by braxex** [comment]
> The api was acting a little funny: throws a 500 when you delete the only plan (probably because it's trying to write 0 plans and that's failing validation). I also saw a 500 when I deleted one plan that had a sanction, so maybe sanction validation had a problem?

Yeah those are all not allowed to be empty arrays right now but we should probably allow them to be written as empty arrays. That would be a good change to make.

**Comment by karla-vm** [comment]
> > The api was acting a little funny: throws a 500 when you delete the only plan (probably because it's trying to write 0 plans and that's failing validation). I also saw a 500 when I deleted one plan that had a sanction, so maybe sanction validation had a problem?
> 
> Yeah those are all not allowed to be empty arrays right now but we should probably allow them to be written as empty arrays. That would be a good change to make.

Is that something I could add to this PR or a separate ticket?

**Review by scwambach — DISMISSED** [review]

**Comment by braxex** [comment]
> > > The api was acting a little funny: throws a 500 when you delete the only plan (probably because it's trying to write 0 plans and that's failing validation). I also saw a 500 when I deleted one plan that had a sanction, so maybe sanction validation had a problem?
> > 
> > 
> > Yeah those are all not allowed to be empty arrays right now but we should probably allow them to be written as empty arrays. That would be a good change to make.
> 
> Is that something I could add to this PR or a separate ticket?
Was a one char change so I just knocked it out. 👍

**Review by karla-vm — COMMENTED** [review]
@braxex thanks for fixing this up! 🥇

**Review by ntsummers1 — APPROVED** [review]

**Review by gmrabian — APPROVED** [review]

**Pull request merged**